### ISSUE TITLE
Issues with Mellanox/InfiniBand (OpenFabrics) device initialization

### ISSUE THREAD

**Issue by DininduSenanayake** [nesi/openmpi-apptainer#1]
```
WARNING: There was an error initializing an OpenFabrics device.

  Local host:   wmc005
  Local device: mlx5_0
```
this means  we're not using the  InfiniBand interconnect optimally, the MPI communication is still working over regular TCP/IP networking  .i.e. application is falling back to TCP-based communication  

1. Add proper RDMA configurations in the container
2. Ensure correct InfiniBand/UCX libraries are installed and configured
3. Possibly bind the relevant host devices and libraries when running the container

**Comment by DininduSenanayake** [comment]
Following recipe should fix it

```bash
Bootstrap: docker
From: ubuntu:22.04

%post
    # Install essential build tools and dependencies
    apt-get update && apt-get install -y --no-install-recommends \
        build-essential \
        ca-certificates \
        wget \
        git \
        tar \
        gzip \
        perl \
        python3 \
        python-is-python3 \
        automake \
        autoconf \
        libtool \
        flex \
        hwloc \
        pkg-config \
        zlib1g-dev \
        libnuma-dev \
        libhwloc-dev \
        libslurm-dev \
        slurm-wlm \
        libpmi2-0 \
        libpmi0 \
        libpmi2-0-dev \
        libpmi0-dev \
        libslurm-dev \
        rdma-core \
        libibverbs1 \
        libibverbs-dev \
        ibverbs-providers \
        librdmacm1 \
        librdmacm-dev \
        perftest \
        libevent-dev \
        libevent-core-2.1-7 \
        libevent-extra-2.1-7 \
        libevent-openssl-2.1-7 \
        libevent-pthreads-2.1-7 \
        libev-dev \
        libssl-dev \
        openssl \
        gfortran \
        && rm -rf /var/lib/apt/lists/*

    # Create IB related symlinks if they don't exist
    mkdir -p /etc/libibverbs.d
    ln -sf /usr/lib/x86_64-linux-gnu/libibverbs.so.1 /usr/lib/libibverbs.so
    ln -sf /usr/lib/x86_64-linux-gnu/librdmacm.so.1 /usr/lib/librdmacm.so

    # Install libevent from source to ensure all components are available
    cd /tmp
    wget https://github.com/libevent/libevent/releases/download/release-2.1.12-stable/libevent-2.1.12-stable.tar.gz
    tar xzf libevent-2.1.12-stable.tar.gz
    cd libevent-2.1.12-stable
    ./configure --prefix=/usr
    make -j$(nproc)
    make install
    ldconfig

    # Create symlinks for PMI libraries
    ln -sf /usr/lib/x86_64-linux-gnu/libpmi2.so.0 /usr/lib/libpmi2.so
    ln -sf /usr/lib/x86_64-linux-gnu/libpmi.so.0 /usr/lib/libpmi.so

    # Install UCX with proper IB support
    export UCX_VERSION=1.14.1
    cd /tmp
    wget https://github.com/openucx/ucx/releases/download/v${UCX_VERSION}/ucx-${UCX_VERSION}.tar.gz
    tar -xvf ucx-${UCX_VERSION}.tar.gz
    cd ucx-${UCX_VERSION}
    ./configure --prefix=/usr \
        --enable-optimizations \
        --enable-mt \
        --enable-numa \
        --with-verbs \
        --with-rdmacm \
        --enable-gdr \
        --enable-mlx5-dv
    make -j$(nproc)
    make install
    ldconfig

    # Install PMIx
    cd /tmp
    wget https://github.com/openpmix/openpmix/releases/download/v4.2.7/pmix-4.2.7.tar.gz
    tar -xvf pmix-4.2.7.tar.gz
    cd pmix-4.2.7
    ./configure --prefix=/usr \
        --disable-python-bindings \
        --with-libevent=/usr \
        --with-libevent-libdir=/usr/lib
    make -j$(nproc)
    make install
    ldconfig

    # Build and install OpenMPI
    cd /tmp
    wget https://download.open-mpi.org/release/open-mpi/v4.1/openmpi-4.1.5.tar.gz
    tar -xvf openmpi-4.1.5.tar.gz
    cd openmpi-4.1.5

    # Configure OpenMPI with updated options for IB support
    ./configure --prefix=/opt/ompi \
        --with-ucx=/usr \
        --with-slurm=/usr \
        --with-pmix=/usr \
        --with-pmi=/usr \
        --with-verbs=/usr \
        --with-verbs-libdir=/usr/lib/x86_64-linux-gnu \
        --enable-mpirun-prefix-by-default \
        --enable-orterun-prefix-by-default \
        --with-pmi-libdir=/usr/lib/x86_64-linux-gnu \
        --enable-mpi1-compatibility \
        --with-hwloc=/usr \
        --enable-builtin-atomics \
        --with-zlib \
        --enable-heterogeneous \
        --enable-mpi-fortran=yes \
        --enable-static=no

    make -j$(nproc)
    make install

    # Cleanup
    cd /
    rm -rf /tmp/ompi /tmp/pmix* /tmp/ucx* /tmp/libevent*

```

**Comment by DininduSenanayake** [comment]
Solved in 9a9590fcc72cecb6187a7533ee48053c2c3801fc
